Change window.open to same tabe instead of new tab

Momcilo42 opened the pull request:
Change the Search module to open results in the same tab.

When using homarr as the default page for a new tab it makes more sense to open search results in the same tab rather than another new tab.
Also added a toggle if someone prefers it to still be in a new tab.

### Category
> One of: Bugfix / Feature

### Overview
> Briefly outline your new changes...
Search module opens results in same tab instead of new tab by default (has toggle)

### New Vars _(if applicable)_
> If you've added any new build scripts, environmental variables, config file options, dependency please outline here.
Added a new variable in the Settings interface to track the toggle switch state.
Added new json files for the toggle text for translations.

ajnart commented:
Just checking this really quick:
To add a translation string you should only add it to the en file (also no need to create one file for a new setting), please remove the unnecessary files and translations 

why not add the toggle as a module option ? 
- you should check out how the ModuleWrapper adds the ModuleOptions and how they are defined in the IModule, you don't need that much code 
(also please not that the translations are dynamically loaded for the module dependant on the module.id field)



Also, I understand there is currently no way of modifying module options on non card modules like docker or searchbar / overseerr / sonarr / radarr but we could add a way , perhaps a context sensitive right click would be nice. Maybe we should add that first. It would also pave the way for the future where there will be no module enabler

ajnart reviewed (changes requested):
please review my comment

manuel-rw commented on src/modules/search/SearchModule.tsx:
```diff
@@ -153,20 +153,22 @@ export default function SearchBar(props: any) {
       }}
       onSubmit={form.onSubmit((values) => {
         const query = values.query.trim();
+        var open_in = config.settings.searchNewTab ?? '_self';
```
Please fix the eslint error:
```Unexpected var, use let or const instead.eslint[no-var](https://eslint.org/docs/rules/no-var)```

manuel-rw replied:
By the way, if you don't see this error, download the Eslint Extension from the Visual Studio Code marketplace:
https://marketplace.visualstudio.com/items?itemName=dbaeumer.vscode-eslint

Momcilo42 replied:
I installed the extension and changed it to const. I don't see any errors now.

manuel-rw replied:
Thank you, you may resolve this issue

manuel-rw commented on src/tools/types.ts:
```diff
@@ -3,6 +3,7 @@ import { OptionValues } from '../modules/ModuleTypes';
 
 export interface Settings {
   searchUrl: string;
+  searchNewTab?: string;
```
The name implies, that it can only have two values:
- It opens in a new tab
- It open in the same tab

Why would you need a string here?

Momcilo42 replied:
The name parameter of window.open is a string so I set it as such. There are some other values besides the two I'm using so I didn't really think about it. Although, I agree with you that the name implies a boolean so I'll make the change.

manuel-rw replied:
Okay, I understand your intention. So, a different question: Will there ever be a different value for this besides ``_self`` and ``_blank``?

Momcilo42 replied:
I'm not sure. In any case it can be a boolean at least for now.

manuel-rw replied:
Okay... For the future, you should read the docs for such questions:
https://developer.mozilla.org/en-US/docs/Web/HTML/Element/a?retiredLocale=de#attr-target

I believe that it wouldn't make any sense to add anything besides ``_self`` and ``_blank``, so I agree and would make it a boolean.

manuel-rw commented on src/tools/translation-namespaces.ts:
```diff
@@ -9,6 +9,7 @@ export const dashboardNamespaces = [
   'settings/general/internationalization',
   'settings/general/module-enabler',
   'settings/general/search-engine',
+  'settings/general/search-new-tab',
```
You don't need to add a new translation namespace. Use the existing search module one instead.

Momcilo42 replied:
I moved it to settings/general/search-engine

manuel-rw commented on src/components/SearchNewTabSwitch/SearchNewTabSwitch.tsx:
```diff
@@ -0,0 +1,62 @@
+    <Group>
+      <div className={classes.root}>
+        <Switch
+          checked={widgetPosition === '_blank'}
```
If you'd use a boolean instead of a string, you wouldn't have to use the equality operator and compare to ``_blank``. See my other comments on the Settings interface.

Momcilo42 replied:
I changed it to boolean as suggested so this is unnecessary.

manuel-rw commented on public/locales/de/settings/general/search-new-tab.json:
```diff
@@ -0,0 +1,3 @@
+{
+    "label": "Open search results in new tab"
```
Don't add any files besides English. Crowdin will automatically detect, that the base language (English) was modified and add the necessary files itself.

Momcilo42 replied:
I removed this file and moved the label to settings/general/search-engine. Just for English, removed the rest.

manuel-rw commented:
> Just checking this really quick: To add a translation string you should only add it to the en file (also no need to create one file for a new setting), please remove the unnecessary files and translations
> 
> why not add the toggle as a module option ?
> 
> * you should check out how the ModuleWrapper adds the ModuleOptions and how they are defined in the IModule, you don't need that much code
>   (also please not that the translations are dynamically loaded for the module dependant on the module.id field)
> 
> Also, I understand there is currently no way of modifying module options on non card modules like docker or searchbar / overseerr / sonarr / radarr but we could add a way , perhaps a context sensitive right click would be nice. Maybe we should add that first. It would also pave the way for the future where there will be no module enabler

I agree with your comment about the module options, but I think that the required changes would be quite big, and it would be better to implement them after the DND update (app shelf rework).

ajnart commented:
> > Just checking this really quick: To add a translation string you should only add it to the en file (also no need to create one file for a new setting), please remove the unnecessary files and translations
> > 
> > why not add the toggle as a module option ?
> > 
> > * you should check out how the ModuleWrapper adds the ModuleOptions and how they are defined in the IModule, you don't need that much code
> >   (also please not that the translations are dynamically loaded for the module dependant on the module.id field)
> > 
> > Also, I understand there is currently no way of modifying module options on non card modules like docker or searchbar / overseerr / sonarr / radarr but we could add a way , perhaps a context sensitive right click would be nice. Maybe we should add that first. It would also pave the way for the future where there will be no module enabler
> 
> I agree with your comment about the module options, but I think that the required changes would take quite some time, and it would be better to implement them after the DND update (app shelf rework).

I was thinking we could show the appropriate module settings on the module enabler by adding a little cog wheel or a contextual right click pop-up, should be easy enough to do

manuel-rw commented:
Hmmm.... Yea, it would make sense. Just keep in mind, that I'll remove parts of the module enabler in the next update

ajnart commented:
> > > > Just checking this really quick: To add a translation string you should only add it to the en file (also no need to create one file for a new setting), please remove the unnecessary files and translations
> > > > why not add the toggle as a module option ?
> > > > 
> > > > * you should check out how the ModuleWrapper adds the ModuleOptions and how they are defined in the IModule, you don't need that much code
> > > >   (also please not that the translations are dynamically loaded for the module dependant on the module.id field)
> > > > 
> > > > Also, I understand there is currently no way of modifying module options on non card modules like docker or searchbar / overseerr / sonarr / radarr but we could add a way , perhaps a context sensitive right click would be nice. Maybe we should add that first. It would also pave the way for the future where there will be no module enabler
> > > 
> > > 
> > > I agree with your comment about the module options, but I think that the required changes would take quite some time, and it would be better to implement them after the DND update (app shelf rework).
> > 
> > 
> > I was thinking we could show the appropriate module settings on the module enabler by adding a little cog wheel or a contextual right click pop-up, should be easy enough to do
> 
> That does sound like a nice change but I feel it's out of the scope of this commit/PR. As it stands now, I put the toggle switch next to the search engine selector as it's the closest thing to it.

The problem is that this feature will probably get overshadowed on the next "big" version release. We can't accept it because it adds too little changes. The next version will completely overall adding services/ modules and interacting with them

Momcilo42 commented:
When will the next "big" version be released? If it's soon I can wait and add this then.
In the meantime, I addressed the comments and made the appropriate changes.

ajnart commented:
I guess we can merge this for 0.10.3 but we'll have to redo it for v0.11 (since a lot of things will change)

what do you think @manuel-rw ?

manuel-rw commented:
Yes, merge it for now. I don't know how much time it will take to complete v0.11, so there might even be a version v0.10.4? - that should be plenty of time.
As @ajnart said, it will probably get refactored in v0.11

manuel-rw commented:
btw, there are outdated threads, which are not resolved. Please resolve them, if you fixed them

Momcilo42 commented:
> Yes, merge it for now. I don't know how much time it will take to complete v0.11, so there might even be a version v0.10.4? - that should be plenty of time. As @ajnart said, it will probably get refactored in v0.11

I resolved all the threads.
I don't care if it gets refactored down the line, as long as the functionality remains.

ajnart commented:
i think we can merge then. Sorry for the resentment towards your PR

manuel-rw commented on public/locales/en/settings/general/search-new-tab.json:
```diff
@@ -0,0 +1,3 @@
+{
+    "label": "Open search results in new tab"
```
Remove old file

Momcilo42 replied:
Removed.

manuel-rw commented:
After this issue has been resolved, and @ajnart approves, we can merge this. Sorry again for the inconvenience and thank you very much for your contribution to v0.10.3!

manuel-rw commented on src/components/SearchNewTabSwitch/SearchNewTabSwitch.tsx:
```diff
@@ -0,0 +1,61 @@
+export function SearchNewTabSwitch() {
+  const { config, setConfig } = useConfig();
+  const { classes, cx } = useStyles();
+  const defaultPosition = config?.settings?.searchNewTab || false;
```
After this, we can merge ✨🚀

Momcilo42 replied:
I never wanted homarr to remain open after i used the search module so in my mind this makes the most sense. It's up to you guys what it should be by default. In this change the switch is right under the Search engine selector so it's not hard to find and should be seen by everyone when they go on to turn on the search module.
If you feel the default should be to open a new tab I'll change it.

ajnart replied:
imo users shouldn't suffer from added features. Not disturbing their usual usage of the app is most important, they'll see the settings and enable it if they want to after updating

Momcilo42 replied:
I see your point. I'll change it now.

Momcilo42 replied:
I had to add the default true value to the defaults file so it could read it on first load.

Momcilo42 commented:
Sorry, I just noticed that he switch state wasn't being read properly after coming back from searching in the same tab, so I updated that.

ajnart commented:
I don't think you need to change [data/configs/default.json](https://github.com/ajnart/homarr/pull/476/files#diff-15f47fecf397171e146573dbbb84eda2fd59fdfa6ab8d5776e4d6fe9593f6218)
The default value should be in the form / code iteself

Momcilo42 commented on data/configs/default.json:
```diff
@@ -10,7 +10,8 @@
   "settings": {
-    "searchUrl": "https://google.com/search?q="
+    "searchUrl": "https://google.com/search?q=",
+    "searchNewTab": true
```
It is if I want the switch to be toggled on by default. If there's no value there it defaults to false, so this seemed to make sense as the searchUrl is also set in the defaults.
I was having lots of issues with the boolean as if the value is null it can also be treated as false. Using ?? fixed the issue.

manuel-rw commented:
@ajnart BUMP

manuel-rw commented:
@ajnart told me that he does not have the time to review this properly. I will review it again after work, so we can merge this for the 0.10.3 release

manuel-rw commented on src/components/SearchNewTabSwitch/SearchNewTabSwitch.tsx:
```diff
@@ -0,0 +1,62 @@
+  const { classes, cx } = useStyles();
+  const defaultPosition = config?.settings?.searchNewTab ?? true;
+  config.settings.searchNewTab = defaultPosition;
+  const [widgetPosition, setWidgetPosition] = useState<boolean>(defaultPosition);
```
``widgetPosition``? Did you copy this name? You should change it to ``openInNewTab`` or something similar

Momcilo42 replied:
Yes, the name is copied, but I thought widget was it was referring to the switch itself. I'll change it.

manuel-rw commented on src/components/SearchNewTabSwitch/SearchNewTabSwitch.tsx:
```diff
@@ -0,0 +1,62 @@
+  const { config, setConfig } = useConfig();
+  const { classes, cx } = useStyles();
+  const defaultPosition = config?.settings?.searchNewTab ?? true;
+  config.settings.searchNewTab = defaultPosition;
```
Is this really necessary?
``??`` is a [logical nullish assignment operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Logical_nullish_assignment), which means, that if searchNewTab is undefined, it will default to ``true``.
Thus, setting the config shouldn't be necessary here.

Momcilo42 replied:
You're right, I must have forgot to remove it after testing. I'll remove it now.

manuel-rw commented:
Looks good. I'll merge now. For any future contributions, please use https://gitmoji.dev/ :)
